akka-stream/src/main/scala/akka/stream/scaladsl/Framing.scala

```diff
@@ -126,65 +127,81 @@ object Framing {
+    val out = Outlet[ByteString]("DelimiterFramingStage.out")
+    override val shape: FlowShape[ByteString, ByteString] = FlowShape(in, out)
 
-    override def onPull(ctx: Context[ByteString]): SyncDirective = doParse(ctx)
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler with OutHandler {
+      private val firstSeparatorByte = separatorBytes.head
+      private var buffer = ByteString.empty
+      private var nextPossibleMatch = 0
 
-    override def onUpstreamFinish(ctx: Context[ByteString]): TerminationDirective =
-      if (buffer.nonEmpty) ctx.absorbTermination()
-      else ctx.finish()
+      override def onPush(): Unit = {
+        buffer ++= grab(in)
+        doParse()
+      }
 
-    private def tryPull(ctx: Context[ByteString]): SyncDirective = {
-      if (ctx.isFinishing) {
-        if (allowTruncation) ctx.pushAndFinish(buffer)
-        else
-          ctx.fail(new FramingException(
-            "Stream finished but there was a truncated final frame in the buffer"))
-      } else ctx.pull()
-    }
+      override def onPull(): Unit = doParse()
 
-    @tailrec
-    private def doParse(ctx: Context[ByteString]): SyncDirective = {
-      val possibleMatchPos = buffer.indexOf(firstSeparatorByte, from = nextPossibleMatch)
-      if (possibleMatchPos > maximumLineBytes)
-        ctx.fail(new FramingException(s"Read ${buffer.size} bytes " +
-          s"which is more than $maximumLineBytes without seeing a line terminator"))
-      else {
-        if (possibleMatchPos == -1) {
-          // No matching character, we need to accumulate more bytes into the buffer
-          nextPossibleMatch = buffer.size
-          tryPull(ctx)
+      override def onUpstreamFinish(): Unit = {
+        if (buffer.isEmpty) {
+          completeStage()
+        } else if (isAvailable(out)) {
+          doParse()
+        } // else swallow the termination and wait for pull
+      }
+
+      private def tryPull(): Unit = {
+        if (isClosed(in)) {
+          if (allowTruncation) {
+            push(out, buffer)
+            completeStage()
+          } else
+            failStage(new FramingException(
+              "Stream finished but there was a truncated final frame in the buffer"))
+        } else pull(in)
+      }
+
+      @tailrec
+      private def doParse(): Unit = {
+        val possibleMatchPos = buffer.indexOf(firstSeparatorByte, from = nextPossibleMatch)
+        if (possibleMatchPos > maximumLineBytes)
+          failStage(new FramingException(s"Read ${buffer.size} bytes " +
+            s"which is more than $maximumLineBytes without seeing a line terminator"))
+        else if (possibleMatchPos == -1) {
+          if (buffer.size > maximumLineBytes)
+            failStage(new FramingException(s"Read ${buffer.size} bytes " +
+              s"which is more than $maximumLineBytes without seeing a line terminator"))
```

Comment:
This is to fix the error reported in #20479. Also slightly flattened nested if/else inside doParse().